xbmc/platform/win32/network/WSDiscoveryWin32.cpp

```diff
@@ -0,0 +1,287 @@
+  {
+    std::wstring addr(address);
+
+    CLog::Log(LOGINFO, "[WS-Discovery]: BYE packet received: device address = '{}'", FromW(addr));
+
+    const std::wstring ip = addr.substr(0, addr.find(L":", 0));
+    auto it = std::find(m_serversIPs.begin(), m_serversIPs.end(), ip);
+
+    // removes server IP from list
+    if (it != m_serversIPs.end())
+    {
+      m_serversIPs.erase(it);
+      CLog::Log(LOGINFO, "[WS-Discovery]: IP '{}' has been removed from the server list.",
+                FromW(ip));
+    }
+  }
+
+  return S_OK;
+}
+
+HRESULT STDMETHODCALLTYPE CClientNotificationSink::SearchFailed(HRESULT hr, LPCWSTR tag)
+{
+  CSingleLock lock(m_criticalSection);
+
+  CLog::Log(LOGINFO,
+            "[WS-Discovery]: The initial search for servers has failed. No servers found.");
+
+  return S_OK;
+}
+
+HRESULT STDMETHODCALLTYPE CClientNotificationSink::SearchComplete(LPCWSTR tag)
+{
+  CSingleLock lock(m_criticalSection);
+
+  CLog::Log(LOGINFO,
+            "[WS-Discovery]: The initial servers search has completed successfully with {} "
+            "server(s) found:",
+            m_serversIPs.size());
+
+  for (const auto& ip : GetServersIPs())
+    CLog::Log(LOGINFO, "    {}", FromW(ip));
+
+  return S_OK;
+}
+
+HRESULT STDMETHODCALLTYPE CClientNotificationSink::QueryInterface(REFIID riid, void** object)
+{
+  if (!object)
+    return E_POINTER;
+
+  *object = nullptr;
+
+  if (__uuidof(IWSDiscoveryProviderNotify) == riid)
+    *object = static_cast<IWSDiscoveryProviderNotify*>(this);
+  else if (__uuidof(IUnknown) == riid)
+    *object = static_cast<IUnknown*>(this);
+  else
+    return E_NOINTERFACE;
+
+  ((LPUNKNOWN)*object)->AddRef();
+
+  return S_OK;
+}
+
+ULONG STDMETHODCALLTYPE CClientNotificationSink::AddRef()
+{
+  ULONG newRefCount = InterlockedIncrement(&m_cRef);
+
+  return newRefCount;
+}
+
+ULONG STDMETHODCALLTYPE CClientNotificationSink::Release()
+{
+  ULONG newRefCount = InterlockedDecrement(&m_cRef);
+
+  if (!newRefCount)
+    delete this;
+
+  return newRefCount;
+}
+
+//==================================================================================
+
+std::shared_ptr<CWSDiscoverySupport> CWSDiscoverySupport::Get()
+{
+  static std::shared_ptr<CWSDiscoverySupport> sWSD(new CWSDiscoverySupport);
+  return sWSD;
+}
+
+CWSDiscoverySupport::CWSDiscoverySupport()
+{
+  Initialize();
+}
+
+CWSDiscoverySupport::~CWSDiscoverySupport()
+{
+  Terminate();
+}
+
+bool CWSDiscoverySupport::Initialize()
+{
+  if (m_initialized)
+    return true;
+
+  if (S_OK == WSDCreateDiscoveryProvider(nullptr, &m_provider))
+  {
+    m_provider->SetAddressFamily(WSDAPI_ADDRESSFAMILY_IPV4);
+
+    if (S_OK == CClientNotificationSink::Create(&m_sink))
+    {
+      if (S_OK == m_provider->Attach(m_sink))
+      {
+        if (S_OK == m_provider->SearchByType(nullptr, nullptr, nullptr, nullptr))
+        {
+          m_initialized = true;
+          CLog::Log(LOGINFO, "[WS-Discovery]: Daemon initialized successfully. Now are listening "
```

Comment:
Any need for the verbosity here? Keep it simple, daemon started. The hello/bye stuff can just be a code comment if needed imo

Reply:
Done!